controllers/helmchart_controller.go

```diff
@@ -579,3 +670,66 @@ func (r *HelmChartReconciler) recordReadiness(chart sourcev1.HelmChart, deleted
 		}, deleted)
 	}
 }
+
+func (r *HelmChartReconciler) indexHelmRepositoryByURL(o runtime.Object) []string {
+	repo, ok := o.(*sourcev1.HelmRepository)
+	if !ok {
+		panic(fmt.Sprintf("Expected a HelmRepository, got %T", o))
+	}
+	u := helm.NormalizeChartRepositoryURL(repo.Spec.URL)
+	if u != "" {
+		return []string{u}
+	}
+	return nil
+}
+
+func (r *HelmChartReconciler) resolveDependencyRepository(ctx context.Context, dep *helmchart.Dependency, namespace string) (*sourcev1.HelmRepository, error) {
+	url := helm.NormalizeChartRepositoryURL(dep.Repository)
+	if url == "" {
+		return nil, fmt.Errorf("invalid repository URL")
+	}
+
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+		client.MatchingField(sourcev1.HelmRepositoryURLIndexKey, url),
+	}
+	var list sourcev1.HelmRepositoryList
+	err := r.Client.List(ctx, &list, listOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve HelmRepositoryList: %w", err)
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, fmt.Errorf("no HelmRepository found")
+}
+
+func (r *HelmChartReconciler) getRepositoryClientOptions(ctx context.Context, repository *sourcev1.HelmRepository) ([]getter.Option, error) {
+	var clientOpts []getter.Option
+
+	if repository.Spec.SecretRef != nil {
+		name := types.NamespacedName{
+			Namespace: repository.GetNamespace(),
+			Name:      repository.Spec.SecretRef.Name,
+		}
+
+		var secret corev1.Secret
+		err := r.Client.Get(ctx, name, &secret)
+		if err != nil {
+			err = fmt.Errorf("auth secret error: %w", err)
+			return nil, err
+		}
+
+		opts, cleanup, err := helm.ClientOptionsFromSecret(secret)
+		if err != nil {
+			err = fmt.Errorf("auth options error: %w", err)
+			return nil, err
+		}
+		defer cleanup()
```

Comment:
I am afraid this will cause Helm to not work for repositories that have TLS credentials configured, as Helm takes a reference to the file(s), and not a copy of the file contents.

I have been wanting to create our own HTTP client, but am currently blocked by the [getter options not being publicly available](https://github.com/helm/helm/blob/ae738d7d87bbd5db7203db7441fb11e32b02df40/pkg/getter/getter.go#L31).

Reply:
Oops, looks like I didn't pay enough attention to the implementation in: `TLSClientConfigFromSecret`.